dashboard/modules/job/job_head.py

```diff
@@ -57,32 +83,81 @@ async def get_package(self,
     async def upload_package(self, req: aiohttp.web.Request):
         package_uri = req.query["package_uri"]
         logger.info(f"Uploading package {package_uri} to the GCS.")
-        upload_package_to_gcs(package_uri, await req.read())
+        try:
+            upload_package_to_gcs(package_uri, await req.read())
+        except Exception:
+            return aiohttp.web.Response(
+                reason=traceback.format_exc().encode("utf-8"),
+                status=aiohttp.web.HTTPInternalServerError.status_code)
 
-        return aiohttp.web.Response()
+        return aiohttp.web.Response(
+            text={"success": True},
+            content_type="application/json",
+            status=aiohttp.web.HTTPOk.status_code,
+        )
 
     @routes.post(JOBS_API_ROUTE_SUBMIT)
     @_ensure_ray_initialized
     async def submit(self, req: aiohttp.web.Request) -> aiohttp.web.Response:
-        # TODO: (jiaodong) Validate if job request is valid without using
-        # pydantic.
-        submit_request = JobSubmitRequest(**(await req.json()))
-        job_id = self._job_manager.submit_job(
-            entrypoint=submit_request.entrypoint,
-            runtime_env=submit_request.runtime_env,
-            metadata=submit_request.metadata)
-
-        resp = JobSubmitResponse(job_id=job_id)
+        result = await self._parse_and_validate_request(req, JobSubmitRequest)
+        # Request parsing failed, returned with Response object.
+        if isinstance(result, aiohttp.web.Response):
+            return result
+        else:
+            submit_request = result
```

Comment:
This is a little bit ugly that we have to duplicate this code everywhere. One alternative would be to do this with a context manager:

```python
with exception_status_code(500):
    # Do stuff, if exception is raised, we return response with 500 and traceback.
```

No need to do it in this PR

Reply:
ahh nice this is definitely better than what we have here. i found it a bit weird towards the end as well but going too far to bother to change it back lol

Reply:
I give that a try and i think current dashboard modules don't really respect exceptions thrown from our own context manager and we can only do it by returning a response but set exception code. So i will leave it as is for now :/